### ISSUE TITLE
Doubts about the meta-learning pipeline

### ISSUE THREAD

**Issue by zhli-cs** [liuyixin-louis/MetaCloak#7]
Thank you for sharing the code. Great work! 

I have a few questions about the meta-learning training pipeline. In the paper, it seems that you make a copy of the model $\theta$ and get $\theta_i$, which is used for K-step lower-level optimization (eq. 6). This K-step optimized model is then used to craft poisoned data (eq. 7). Then, such new poisoned data is utilized to optimize the surrogate model $\theta$ (eq. 8). 

However, in your codes, the iteration starting from line 578 to line 606 seems to use the same model "f" continuously for both K-step unrolling optimization and noise crafting. I did not find the corresponding training steps to optimize the original surrogate model with newly crafted poisoned data (i.e., eq. 8). Could you please help me with this issue? 

![Image](https://github.com/user-attachments/assets/9044360a-cb4c-4ccd-9a63-364bf8787e50)

![Image](https://github.com/user-attachments/assets/fe1f438b-637b-4561-a8f4-ba58e53069c3)

I'm not sure if my understanding is correct or not. If not, it would be great if you could provide me with the correct understanding of this example.

Thanks a lot for your help.

**Comment by liuyixin-louis** [comment]
Thanks for the careful reading and the detailed question!

You're right that the code structure looks different from the pseudocode at first glance, but the core bi-level optimization logic is preserved. Let me clarify:

**1. `train_few_step` uses `copy.deepcopy` internally (line 60)**

```python
unet, text_encoder = copy.deepcopy(models[0]), copy.deepcopy(models[1])
```

So when `train_few_step(f, ...)` is called, the original `f` is not modified in-place — a deep copy is created, trained for K steps, and returned. This ensures the perturbation crafting step (`defender.perturb`) and the inner-loop training step operate on the correct model states.

**2. The alternating loop implements the bi-level optimization**

The loop structure:
```python
for j in range(args.interval // args.advance_steps):
    perturbed_data = defender.perturb(f, perturbed_data, ...)   # eq. 7 & 8: craft/update perturbation
    f = train_few_step(f, perturbed_data, advance_steps)         # eq. 6: K-step inner-loop on new perturbation
```

- `defender.perturb()` corresponds to **eq. 7 & 8** — computing the generation loss with the current surrogate and updating the perturbation via gradient ascent.
- `train_few_step()` corresponds to **eq. 6** — training a fresh copy of the surrogate on the updated perturbation for K steps.

So eq. 8 (the perturbation update) is implemented inside `defender.perturb()`, and the surrogate is subsequently retrained on the newly crafted perturbation in the next `train_few_step` call.

**3. First-order approximation**

As noted in the paper ("Following MAML, a first-order gradient approximation is used for efficient perturbation crafting"), the code uses a FOMAML-style simplification where we alternate between perturbation crafting and surrogate training sequentially, rather than maintaining a strict θ/θ_i separation with full second-order gradients. This is a well-established simplification in meta-learning that works comparably in practice.

Hope this clarifies! Feel free to follow up if you have further questions.

**Comment by liuyixin-louis** [comment]
Closing as addressed above. Feel free to reopen if needed.
